### PR TITLE
Load the full dependency tree deterministically during VMR sync

### PR DESCRIPTION
**Pull request by premun**
This code change preloads the full repository tree to figure out which versions of repositories it should ingest.

Tests will be added once we have the other E2E infra in place.

Resolves https://github.com/dotnet/arcade/issues/11725
Resolves https://github.com/dotnet/arcade/issues/11818

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/DarcLib/Helpers/IGitFileManager.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using System.Xml;
+using Newtonsoft.Json.Linq;
+using NuGet.Versioning;
+
+namespace Microsoft.DotNet.DarcLib;
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/DarcLib/Helpers/IGitFileManager.cs:12]:
```suggestion
namespace Microsoft.DotNet.DarcLib;

```

File: src/Microsoft.DotNet.Darc/src/DarcLib/Helpers/IGitFileManager.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+public interface IGitFileManager
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/DarcLib/Helpers/IGitFileManager.cs:13]: This is a public interface; do we want to add some meaningful description?

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManagerBase.cs [hunk 3]
```diff
@@ -60,35 +69,75 @@ protected void Commit(string commitMessage, Signature author)
...
+            { root.Mapping, root },
+        };
 
-        var dependencies = _versionDetailsParser.ParseVersionDetailsXml(versionDetailsContent, true)
-            .Where(dep => dep.SourceBuild is not null);
+        var reposToScan = new Queue<DependencyUpdate>();
+        reposToScan.Enqueue(transitiveDependencies.Values.Single());
 
-        var result = new List<(DependencyDetail, SourceMapping)>();
+        _logger.LogInformation("Finding transitive dependencies for {mapping}:{revision}..", root.Mapping.Name, root.TargetRevision);
```
Comment (andriipatsula) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManagerBase.cs:84]: I am not sure if a `.` (dot symbol) could be part of revision.. what about to put `{mapping}:{revision}` into quotes? Cause if revision could have a dot symbol, then `{mapping}:{revision}..` might have `..` or `...` at the end.

**Review by oleksandr-didyk — APPROVED** [review]

**Review by andriipatsula — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManagerBase.cs (hunk 3 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrManagerBase.cs:84]: Valid point, however revision here is the commit SHA so we're fine

**Review by andriipatsula — APPROVED** [review]

**Pull request merged**